sentry-ruby/lib/sentry/profiler.rb

```diff
@@ -0,0 +1,228 @@
+      @in_app_pattern = Regexp.new("^(#{@project_root}/)?#{@app_dirs_pattern}")
+    end
+
+    def start
+      return unless @sampled
+
+      @started = StackProf.start(interval: DEFAULT_INTERVAL,
+                                 mode: :wall,
+                                 raw: true,
+                                 aggregate: false)
+
+      @started ? log('Started') : log('Not started since running elsewhere')
+    end
+
+    def stop
+      return unless @sampled
+      return unless @started
+
+      StackProf.stop
+      log('Stopped')
+    end
+
+    def profile_context
+      { profile_id: @event_id }
+    end
+
+    # Sets initial sampling decision of the profile.
+    # @return [void]
+    def set_initial_sample_decision(transaction_sampled)
+      unless @profiling_enabled
+        @sampled = false
+        return
+      end
+
+      unless transaction_sampled
+        @sampled = false
+        log('Discarding profile because transaction not sampled')
+        return
+      end
+
+      case @profiles_sample_rate
+      when 0.0
+        @sampled = false
+        log('Discarding profile because sample_rate is 0')
+        return
+      when 1.0
+        @sampled = true
+        return
+      else
+        @sampled = Random.rand < @profiles_sample_rate
+      end
+
+      log('Discarding profile due to sampling decision') unless @sampled
+    end
+
+    def to_hash
+      return nil unless @sampled
+      return nil unless @started
+
+      results = StackProf.results
+      return nil unless results
+      return nil if results.empty?
+      return nil if results[:samples] == 0
+      return nil unless results[:raw]
+
+      frame_map = {}
+
+      frames = results[:frames].to_enum.with_index.map do |frame, idx|
+        frame_id, frame_data = frame
+
+        # need to map over stackprof frame ids to ours
+        frame_map[frame_id] = idx
+
+        in_app = in_app?(frame_data[:file])
+        filename = compute_filename(frame_data[:file], in_app)
+        function, mod = split_module(frame_data[:name])
+
+        frame_hash = {
+          abs_path: frame_data[:file],
+          function: function,
+          filename: filename,
+          in_app: in_app
+        }
+
+        frame_hash[:module] = mod if mod
+        frame_hash[:lineno] = frame_data[:line] if frame_data[:line]
+
+        frame_hash
+      end
+
+      idx = 0
+      stacks = []
+      num_seen = []
+
+      # extract stacks from raw
+      # raw is a single array of [.., len_stack, *stack_frames(len_stack), num_stack_seen , ..]
+      while (len = results[:raw][idx])
+        idx += 1
+
+        # our call graph is reversed
+        stack = results[:raw].slice(idx, len).map { |id| frame_map[id] }.compact.reverse
+        stacks << stack
+
+        num_seen << results[:raw][idx + len]
+        idx += len + 1
+
+        log('Unknown frame in stack') if stack.size != len
+      end
```

Comment:
So if I'm understanding this correctly if the profile looks like

```
at t = 0 stack is a -> b
at t = 1 stack is a -> b -> c
at t = 2 stack is a -> b
```

the `stacks` lists will contain `a -> b` twice since it only merges identical stacks that are adjacent to each other?

This could potentially result in a massive payload from the duplicate stacks for certain types of profiles, but not a blocker for this initial release.

Reply:
yes, this only optimizes adjacent samples, logic is below
https://github.com/tmm1/stackprof/blob/52d1df6502eaede5061a2b5d28026e982bcf262a/ext/stackprof/stackprof.c#L513-L528

If we write our own extension later, I'll keep this in mind to also optimize.